src/libraries/Microsoft.Extensions.Configuration.Binder/gen/CollectionSpec.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
+{
+    internal abstract record CollectionSpec : TypeSpec
+    {
+        public CollectionSpec(ITypeSymbol type) : base(type)
+        {
+            IsReadOnly = type.IsReadOnly;
+            IsInterface = type is INamedTypeSymbol { TypeKind: TypeKind.Interface };
+        }
+
+        public required TypeSpec ElementType { get; init; }
+
+        public bool IsReadOnly { get; }
+
+        public bool IsInterface { get; }
+
+        public CollectionSpec? ConcreteType { get; init; }
+    }
+
+    internal sealed record EnumerableSpec : CollectionSpec
+    {
+        public EnumerableSpec(ITypeSymbol type) : base(type) { }
+
+        public override TypeSpecKind SpecKind { get; init; } = TypeSpecKind.Enumerable;
+    }
+
+    internal sealed record DictionarySpec : CollectionSpec
+    {
+        public DictionarySpec(INamedTypeSymbol type) : base(type) { }
+
+        public override TypeSpecKind SpecKind => TypeSpecKind.Dictionary;
```

Comment:
nit: Should this be `{ get; init; }` as EnumerableSpec?

Reply:
It depends: if it is converted into an auto-property then it will be part of the record's equality comparison implementation.

FWIW the equality comparison generated for C# records does take polymorphism into account, so you wouldn't need to take the "kind" into account:

https://sharplab.io/#v2:D4AQTAjAsAUCDMACciDCiDetE+UgTgKYDGA9vgCaIBCAhgM6EAUAlgHYAuiAHgJQDc2XEJwJERMpUQARQvhYA3QhVaceAGkTsuAT16IAXDQbM+mAL6xzQA==